File: src/76.ts
// Computes the number of ways `target` can be written as a sum of at least two positive integers.
//
// See {@link https://projecteuler.net/problem=76}
export default function compute(target: number) {
  return partition(1, target) - 1;
}

const map = new Map<string, number>();

function partition(k: number, n: number): number {
  if (k > n) {
    return 0;
  }

  if (k === n) {
    return 1;
  }

  const key = `${k},${n}`;
  if (map.has(key)) {
    return map.get(key)!;
  }

  const value = partition(k + 1, n) + partition(k, n - k);
  map.set(key, value);
  return value;
}

File: project_euler/76.py
# Problem 76: Counting Summations
#
# How many different ways can one hundred be written as a sum of at least two positive integers?
#
# See https://projecteuler.net/problem=76
from functools import cache


@cache
def partition(k: int, n: int) -> int:
    if k > n:
        return 0
    if k == n:
        return 1
    return partition(k + 1, n) + partition(k, n - k)


def run(target=100) -> int:
    return partition(1, target) - 1


def test_run():
    assert run() == 190569291
